pkg/sql/create_stats.go

```diff
@@ -293,44 +297,80 @@ func createStatsDefaultColumns(
 
 	requestedStats := make(map[string]struct{})
 
+	// trackStatsIfNotExists adds the given column IDs as a set to the
+	// requestedStats set. If the columnIDs were not already in the set, it
+	// returns true.
+	trackStatsIfNotExists := func(colIDs []sqlbase.ColumnID) bool {
+		key := makeColStatKey(colIDs)
+		if _, ok := requestedStats[key]; ok {
+			return false
+		}
+		requestedStats[key] = struct{}{}
+		return true
+	}
+
+	// addSingleColumnHistogramIfNotExists appends histogram column stats for
+	// the given column ID if the colStats if it does not already exist.
+	addSingleColumnHistogramIfNotExists := func(colID sqlbase.ColumnID) {
+		colList := []sqlbase.ColumnID{colID}
+
+		// Check for existing stats and remember the requested stats.
+		if !trackStatsIfNotExists(colList) {
+			return
+		}
+
+		colStats = append(colStats, jobspb.CreateStatsDetails_ColStat{
+			ColumnIDs:    colList,
+			HasHistogram: true,
+		})
+	}
+
 	// Add column stats for the primary key.
 	for i := range desc.PrimaryIndex.ColumnIDs {
-		if i != 0 && !multiColEnabled {
-			break
+		// Add a histogram for each column in the primary key.
+		addSingleColumnHistogramIfNotExists(desc.PrimaryIndex.ColumnIDs[i])
+
+		// Only collect multi-column stats if enabled.
+		if i == 0 || !multiColEnabled {
+			continue
 		}
 
 		colIDs := desc.PrimaryIndex.ColumnIDs[: i+1 : i+1]
 
 		// Remember the requested stats so we don't request duplicates.
-		key := makeColStatKey(colIDs)
-		requestedStats[key] = struct{}{}
+		trackStatsIfNotExists(colIDs)
 
+		// Only generate non-histogram multi-column stats.
 		colStats = append(colStats, jobspb.CreateStatsDetails_ColStat{
 			ColumnIDs:    colIDs,
-			HasHistogram: i == 0,
+			HasHistogram: false,
 		})
 	}
 
 	// Add column stats for each secondary index.
 	for i := range desc.Indexes {
 		for j := range desc.Indexes[i].ColumnIDs {
+			// Generate a histogram for each column indexed in forward indexes.
+			if desc.Indexes[i].Type == sqlbase.IndexDescriptor_FORWARD {
+				addSingleColumnHistogramIfNotExists(desc.Indexes[i].ColumnIDs[j])
+			}
+
 			if j != 0 && !multiColEnabled {
-				break
+				continue
 			}
 
 			colIDs := desc.Indexes[i].ColumnIDs[: j+1 : j+1]
 
 			// Check for existing stats and remember the requested stats.
-			key := makeColStatKey(colIDs)
-			if _, ok := requestedStats[key]; ok {
+			if !trackStatsIfNotExists(colIDs) {
 				continue
 			}
-			requestedStats[key] = struct{}{}
 
-			// Only generate a histogram for forward indexes.
+			// Generate non-histogram stats for the first column of an inverted
+			// index or for multi-column stats.
```

Comment:
This is a bit confusing, and I don't love it. It's a bit too clever. It's not super clear that we get to this point when `j == 0` and the index is an inverted index. I couldn't figure out anything simpler besides two entirely separate paths for forward vs. inverted with quite a bit of duplication. I'm to ideas here.